File: py/binaryheap.py

import math

class BinaryHeap:
    def __init__(self):
        self.items = []

    def top(self):
        return self.items[0]
    
    def swim(self, startIndex):
        if startIndex == 0:
            return
        parentIndex = self.getParentIndex(startIndex)
        if self.items[startIndex] > self.items[parentIndex]:
            self.items[parentIndex], self.items[startIndex] = self.items[startIndex], self.items[parentIndex]
            self.swim(parentIndex)
        
    def sink(self, startIndex):
        if len(self.items) < 2:
            return
        leftIndex = self.leftChildIndex(startIndex)
        rightIndex = self.rightChildIndex(startIndex)
        if leftIndex >= len(self.items) and rightIndex >= len(self.items):
            return
        if leftIndex >= len(self.items):
            swapIndex = rightIndex
        if rightIndex >= len(self.items):
            swapIndex = leftIndex
        else:
            swapIndex = self.items[leftIndex] > self.items[rightIndex] if leftIndex else rightIndex
        if self.items[startIndex] < self.items[swapIndex]:
            self.items[startIndex], self.items[swapIndex] = self.items[swapIndex], self.items[startIndex]
            self.sink(swapIndex)

    def leftChildIndex(self, index):
        return math.floor((index + 1 * 2) - 1)

    def rightChildIndex(self, index):
        return math.floor((index + 1) * 2)

    def getParentIndex(self, index):
        return math.floor((index - 1) / 2)

    def pop(self):
        self.items[0], self.items[-1] = self.items[-1], self.items[0]
        del self.items[-1]
        self.sink(0)

    def add(self, data):
        self.items.append(data)
        self.swim(len(self.items)-1)


bh = BinaryHeap()

bh.add(3)
bh.add(5)
print(bh.top())
bh.add(10)
bh.pop()
print(bh.top())

File: js/binaryheap.js

class BinaryHeap {
    constructor() {
        this.items = [];
        this.size = 0;
    }

    add(data) {
        this.items.push(data);
        this.swim(this.size);
        this.size++;
    }

    swim(index) {
        // parent and swap and recurse
        const parentIndex = this.getParentIndex(index);
        if (this.items[index] > this.items[this.getParentIndex(index)]) {
            [this.items[index], this.items[this.getParentIndex(index)]] = [this.items[this.getParentIndex(index)], this.items[index]];
            this.swim(parentIndex);
        }
    }

    getParentIndex(index) {
        return Math.floor((index - 1) / 2);
    }

    getLeftChildIndex(index) {
        return Math.floor((index + 1 * 2) - 1);
    }

    getRightChildIndex(index) {
        return Math.floor((index + 1) * 2);
    }

    pop() {
        [this.items[0], this.items[this.size-1]] = [this.items[this.size-1], this.items[0]];
        this.items.pop();
        this.size--;
        this.sink(0);
    }

    sink(index) {
        // compare left and right to get swapIndex, then check for swap and recurse
        let biggerChildIndex;
        let left = this.items[this.getLeftChildIndex(index)];
        let right = this.items[this.getRightChildIndex(index)];
        if (left == undefined && right == undefined) {
            return;
        }
        if (right == undefined) {
            biggerChildIndex = this.getLeftChildIndex(index);
        } 
        else {
            biggerChildIndex = this.items[this.getLeftChildIndex(index)] > this.items[this.getRightChildIndex(index)] ? this.getLeftChildIndex(index) : this.getRightChildIndex(index);
        }    
        if (this.items[index] < this.items[biggerChildIndex]) {
            [this.items[index], this.items[biggerChildIndex]] = [this.items[biggerChildIndex], this.items[index]];
            this.sink(biggerChildIndex);
        }
    }

    top() {
        return this.items[0];
    }

}

bh = new BinaryHeap();
bh.add(3);
bh.add(5);
bh.add(2);
bh.add(7);
bh.pop();
bh.top();
bh.pop();
bh.add(1);
